### PR TITLE
[sled-agent] Store service configuration data in duplicate in M.2s

### PR DESCRIPTION
**Pull request by smklein**
- Creates a `Ledger` structure which makes it easy to write toml-serializable data to and from M.2s.
- Uses this `Ledger` structure to store all service configuration information in duplicate on the M.2s.

Fixes https://github.com/oxidecomputer/omicron/issues/2969

### CODE REVIEW

**Comment by smklein** [comment]
FWIW, I've tested this on `BRM42220026` in rack2, and the configs aren't in `/var/oxide` anymore.

I *am* seeing them in `/pool/int/0d8f680f-0907-4170-822b-5c49d43a7660/config/` and `/pool/int/2acb2cf2-ed09-4009-b4ce-3b651552e166/config/` now.

**Review by andrewjstone — COMMENTED** [review]
@smklein Nice work! I like the approach taken very much.

Unfortunately, I have one somewhat major concern. I'm not sure how practical of a concern it is as it depends on hardware behavior. There is the possibility of data loss for a ledger in the following scenario:

1. Generation 1 is written to both M.2s( `A` and `B`)
2. Generation 2 is written only to `A` - `B` fails to write
3. Sled-agent reboots
4. `Ledger::new` reads from `B`, but fails to read from `A`
5. The ledger now points, incorrectly, to Generation 1.

At this point we have data loss, but things can get even more confusing, depending upon the failure modes of the M.2s. Let's say we continue with the following steps:

6. Generation 2 (with different data) is written to A.
7. Sled-agent reboots
8. `Ledger::new` reads both versions at generation 2 and picks `A` given its current logic as long as `A` is first in the path list:

```rust
// Return the ledger with the highest generation number.
        let ledger = ledgers.into_iter().reduce(|prior, ledger| {
            if ledger.is_newer_than(&prior) {
                ledger
            } else {
                prior
            }
        });
```

This problem is baked into the fact that you can't do consensus with only 2 nodes if those nodes can fail in arbitrary ways. Whether the M.2s can fail in arbitrary ways is unknown to me, but I'd really like to preclude the possibility of doing the wrong thing without relying on hardware behavior if at all possible. I can see two possible ways of going about resolving this issue:

1. If we ever fail to read or write from an M.2 we refuse to ever bring that M.2 back online. In short we tolerate the failure by making it permanent. 
2. We do not bump ledger generation numbers in sled-agent, but instead bump them in Nexus, along with saving the latest ledger configuration in Nexus. We then would know if we read back stale data, which we could go ahead and rewrite based on what was put in CockroachDB via Nexus.

It's unclear to me if either of these is actually feasible, as presumably the zones must come up in order to be able to talk to Nexus in the first place. However, maybe *updates* can go to nexus as in option 2.

CC @rmustacc

**Comment by smklein** [comment]
> This problem is baked into the fact that you can't do consensus with only 2 nodes if those nodes can fail in arbitrary ways. Whether the M.2s can fail in arbitrary ways is unknown to me, but I'd really like to preclude the possibility of doing the wrong thing without relying on hardware behavior if at all possible. I can see two possible ways of going about resolving this issue:

So, first of all, I totally agree with you about this mismatch being possible. In a longer-term plan, I would like for Nexus to be able to send the request for services / datasets to the sled as:

"Here are *all* the services you should run, with a generation number"

(I've updated https://github.com/oxidecomputer/omicron/issues/732 to include this implementation detail)

Such an API means that this ledger becomes a cache of data that's stored in CRDB, and which can get updated when Nexus comes online.

> 2. We do not bump ledger generation numbers in sled-agent, but instead bump them in Nexus, along with saving the latest ledger configuration in Nexus. We then would know if we read back stale data, which we could go ahead and rewrite based on what was put in CockroachDB via Nexus.

I 100% think this is feasible, and is something we should do. It's easier for "non-dataset services" than "dataset services" due to the existing shape of the internal API, but I think both can be vectorized + generation'd. 

I've created https://github.com/oxidecomputer/omicron/issues/2977  and https://github.com/oxidecomputer/omicron/issues/2978 -- sub-issues of #732 -- for us to track.

**Comment by andrewjstone** [comment]
> I 100% think this is feasible, and is something we should do. It's easier for "non-dataset services" than "dataset services" due to the existing shape of the internal API, but I think both can be vectorized + generation'd.

> I've created https://github.com/oxidecomputer/omicron/issues/2977 and https://github.com/oxidecomputer/omicron/issues/2978 -- sub-issues of https://github.com/oxidecomputer/omicron/issues/732 -- for us to track.

Awesome! Thank you!

**Review by andrewjstone — APPROVED** [review]
Based on the understanding around Nexus issuing updates for services when online and the Ledger acting as a cache I think we should go ahead and merge this in once the test bug is fixed.

File: sled-agent/src/ledger.rs [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+        Ok(())
+    }
+}
+
+#[async_trait]
+pub trait Ledgerable:
+    Default + DeserializeOwned + Serialize + Send + Sync
+{
+    /// Returns true if [Self] is newer than `other`.
+    fn is_newer_than(&self, other: &Self) -> bool;
```
[the remark this answers is not in the document]
Comment (smklein) [reply sled-agent/src/ledger.rs:150]: I think I care more about the explicitness here. I *only* care about the generation number, and do not want to impose this constraint on any other aspect of the structure. I also don't *think* I care about Equality at this point?

File: sled-agent/src/ledger.rs [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+            return Err(Error::FailedToAccessStorage);
+        }
+        Ok(())
+    }
+
+    // Atomically serialize and write the ledger to storage.
+    //
+    // We accomplish this by first writing to a temporary file, then
+    // renaming to the target location.
+    async fn atomic_write(&self, path: &Path) -> Result<(), Error> {
```
[the remark this answers is not in the document]
Comment (smklein) [reply sled-agent/src/ledger.rs:124]: I'm going to try to split this conversion out in a separate change - I think it's a little inconsistent to apply it to exclusively this PR.

File: sled-agent/src/ledger.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewjstone) [reply sled-agent/src/ledger.rs:150]: Works for me.

File: sled-agent/src/ledger.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrewjstone) [reply sled-agent/src/ledger.rs:124]: Works for me.

[the remark this answers is not in the document]
Comment (smklein) [reply sled-agent/src/ledger.rs:124]: The diff is getting messed up, but https://github.com/oxidecomputer/omicron/pull/2988 works through this

File: sled-agent/src/ledger.rs [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+        ledger.data_mut().contents = "new contents".to_string();
+        ledger.commit().await.expect("Failed to write ledger");
+
+        assert!(config_paths[0].exists());
+        assert!(config_paths[1].exists());
+
+        drop(ledger);
+
+        // Let's write again, but only using one of the two config dirs.
+        let mut ledger = Ledger::<Data>::new(&log, config_paths[..=1].to_vec())
```
[the remark this answers is not in the document]
Comment (smklein) [reply sled-agent/src/ledger.rs]: good catch, fixed in ef6756d10be82a0c9fea1df597989acccaee1260

**Pull request merged**